mesonbuild/interpreter/type_checking.py

```diff
@@ -448,3 +461,281 @@ def link_whole_validator(values: T.List[T.Union[StaticLibrary, CustomTarget, Cus
+            return 'must be X.Y.Z where Y is [0, 255] and Y, Z are optional'
+        if len(parts) == 3 and int(parts[2]) > 255:
+            return 'must be X.Y.Z where Z is [0, 255] and Y, Z are optional'
+    return None
+
+def _convert_darwin_versions(val: T.List[T.Union[str, int]]) -> T.Optional[T.Tuple[str, str]]:
+    if not val:
+        return None
+    elif len(val) == 1:
+        v = str(val[0])
+        return (v, v)
+    return (str(val[0]), str(val[1]))
+
+_EXCLUSIVE_SHARED_LIB_KWS: T.List[KwargInfo] = [
+    _VS_MODULE_DEF_KW,
+    _LINK_WITH_KW,
+    KwargInfo('version', (str, NoneType)),
+    KwargInfo('soversion', (str, int, NoneType), convertor=lambda x: str(x) if x is not None else None),
+    KwargInfo(
+        'darwin_versions',
+        ContainerTypeInfo(list, (str, int)),
+        default=[],
+        listify=True,
+        validator=_validate_darwin_versions,
+        convertor=_convert_darwin_versions,
+        since='0.48.0',
+    ),
+]
+
+SHARED_LIB_KWS: T.List[KwargInfo] = (
+    _BUILD_TARGET_KWS + _LANGUAGE_KWS + _EXCLUSIVE_SHARED_LIB_KWS +
+    [_RUST_CRATE_TYPE_KW.evolve(
+        validator=in_set_validator({'lib', 'dylib', 'cdylib', 'proc-macro'}),
+        since_values={'proc-macro': '0.62.0'},
+    )]
+)
+
+SHARED_MOD_KWS: T.List[KwargInfo] = [
+    *_BUILD_TARGET_KWS,
+    *_LANGUAGE_KWS,
+    _VS_MODULE_DEF_KW.evolve(since='0.52.0'),
+    # Shared modules can additionally by linked with Executables
+    KwargInfo(
+        'link_with',
+        ContainerTypeInfo(
+            list,
+            (BothLibraries, SharedLibrary, StaticLibrary,
+             SharedModule, CustomTarget, CustomTargetIndex,
+             Executable)),
+        default=[],
+        listify=True,
+    ),
+]
+
+BOTH_LIB_KWS: T.List[KwargInfo] = (
+    _BUILD_TARGET_KWS + _LANGUAGE_KWS + _EXCLUSIVE_SHARED_LIB_KWS + _EXCLUSIVE_STATIC_LIB_KWS +
+    # XXX: rust_crate_type (and rust in general?) is busted with both lib
+    [
+        _RUST_CRATE_TYPE_KW.evolve(
+            validator=in_set_validator({'lib', 'dylib', 'cdylib', 'rlib', 'staticlib', 'proc-macro'}),
+            since_values={'proc-macro': '0.62.0'}),
+        _PIE_KW,
+    ]
+)
+
+def _win_subsystem_validator(value: T.Optional[str]) -> T.Optional[str]:
+    value = value.lower()
+    if re.fullmatch(r'(boot_application|console|efi_application|efi_boot_service_driver|efi_rom|efi_runtime_driver|native|posix|windows)(,\d+(\.\d+)?)?', value) is None:
+        return f'Invalid value for win_subsystem: {value}.'
+    return None
+
+
+_EXCLUSIVE_EXECUTABLE_KWS: T.List[KwargInfo] = [
+    KwargInfo('export_dynamic', bool, default=False, since='0.45.0'),
+    KwargInfo('gui_app', (bool, NoneType), deprecated='0.56.0', deprecated_message="Use 'win_subsystem' instead."),
+    KwargInfo('implib', (bool, str, NoneType), since='0.42.0'),
+    KwargInfo('win_subsystem', (str, NoneType), since='0.56.0', validator=_win_subsystem_validator),
```

Comment:
hmm... I guess we could a deprecated version in non executable space? though that might require some care to avoid warnings in the case of calling `build_target` directly...